Optimize `Wtf8Buf::into_string` for the case where it contains UTF-8.

sunfishcode opened the pull request:
Add a `is_known_utf8` flag to `Wtf8Buf`, which tracks whether the
string is known to contain UTF-8. This is efficiently computed in many
common situations, such as when a `Wtf8Buf` is constructed from a `String`
or `&str`, or with `Wtf8Buf::from_wide` which is already doing UTF-16
decoding and already checking for surrogates.

This makes `OsString::into_string` O(1) rather than O(N) on Windows in
common cases.

And, it eliminates the need to scan through the string for surrogates in
`Args::next` and `Vars::next`, because the strings are already being
translated with `Wtf8Buf::from_wide`.

Many things on Windows construct `OsString`s with `Wtf8Buf::from_wide`,
such as `DirEntry::file_name` and `fs::read_link`, so with this patch,
users of those functions can subsequently call `.into_string()` without
paying for an extra scan through the string for surrogates.

r? @ghost

sunfishcode committed:
Optimize `Wtf8Buf::into_string` for the case where it contains UTF-8.

Add a `is_known_utf8` flag to `Wtf8Buf`, which tracks whether the
string is known to contain UTF-8. This is efficiently computed in many
common situations, such as when a `Wtf8Buf` is constructed from a `String`
or `&str`, or with `Wtf8Buf::from_wide` which is already doing UTF-16
decoding and already checking for surrogates.

This makes `OsString::into_string` O(1) rather than O(N) on Windows in
common cases.

And, it eliminates the need to scan through the string for surrogates in
`Args::next` and `Vars::next`, because the strings are already being
translated with `Wtf8Buf::from_wide`.

Many things on Windows construct `OsString`s with `Wtf8Buf::from_wide`,
such as `DirEntry::file_name` and `fs::read_link`, so with this patch,
users of those functions can subsequently call `.into_string()` without
paying for an extra scan through the string for surrogates.

jyn514 commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -195,15 +222,17 @@ impl Wtf8Buf {
-                    string.push_code_point_unchecked(code_point)
+                    string.push_code_point_unchecked(code_point);
+                    // The string now contains an unpaired surrogate.
+                    string.is_known_utf8 = false;
```
Can `push_code_point_unchecked` panic? If so, we should probably move the `is_known_utf8` assignment to the line above.

sunfishcode replied:
It can at least oom, so I've now moved the `is_known_utf8` assignment to the line above.

jyn514 commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -313,7 +345,15 @@ impl Wtf8Buf {
+                // If we're pushing a string containing a surrogate, we may no
+                // longer have UTF-8.
+                if other.next_surrogate(0).is_some() {
+                    self.is_known_utf8 = false;
```
ditto here for panic safety

jyn514 commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -575,6 +627,11 @@ impl Wtf8 {
 
+    /// Creates an owned `Wtf8Buf` from a borrowed `Wtf8`.
+    pub fn to_owned(&self) -> Wtf8Buf {
+        Wtf8Buf { bytes: self.bytes.to_vec(), is_known_utf8: false }
```
```suggestion
        Wtf8Buf { bytes: self.bytes.to_vec(), is_known_utf8: self.is_known_utf8 }
```

sunfishcode replied:
I only added `is_known_utf8` to `Wtf8Buf`, and not `Wtf`, because `Wtf` does `transmute` tricks with `&[u8]`.

If it were important to set `is_known_utf8` here, we could do a scan here. But since the main purpose of `is_known_utf8` is to avoid doing a scan later, doing a scan eagerly in order to set it isn't a clear win.

jyn514 replied:
I'm confused. Why is scanning here necessary? There's no transformation happening here, it should always be the same as whatever `is_known_utf8` is currently set to.

sunfishcode replied:
This is in `Wtf8::to_owned`, and `Wtf8` doesn't have a `is_known_utf8` field itself, and it doesn't always point at a `Wtf8Buf`. It may be pointing at an arbitrary `&[u8]` where there is no `is_known_utf8` field.

jyn514 commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -704,12 +762,18 @@ impl Wtf8 {
-        Wtf8Buf { bytes: self.bytes.to_ascii_lowercase() }
+        Wtf8Buf {
+            bytes: self.bytes.to_ascii_lowercase(),
+            is_known_utf8: self.next_surrogate(0).is_none(),
```
does it make sense to scan eagerly? maybe this should just be `false` instead?

sunfishcode replied:
Makes sense.

sunfishcode committed:
Panic safety.

sunfishcode committed:
Don't eagerly scan for `is_known_utf8` in `to_ascii_lowercase`/`uppercase`.

sunfishcode committed:
Remove `is_known_utf8` checks from more tests where it's no longer set.

sunfishcode commented:
r? rust-lang/libs

joshtriplett commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit f4c93bcbe837ab5220beaa2892fc249f49a9e740 has been approved by `joshtriplett`

<!-- @bors r=joshtriplett f4c93bcbe837ab5220beaa2892fc249f49a9e740 -->
<!-- homu: {"type":"Approved","sha":"f4c93bcbe837ab5220beaa2892fc249f49a9e740","approver":"joshtriplett"} -->

bors commented:
:hourglass: Testing commit f4c93bcbe837ab5220beaa2892fc249f49a9e740 with merge 7f13587a59ff3c5350b77bf9334994f91f05d9fc...
<!-- homu: {"type":"BuildStarted","head_sha":"f4c93bcbe837ab5220beaa2892fc249f49a9e740","merge_sha":"7f13587a59ff3c5350b77bf9334994f91f05d9fc"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6480020907?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6480020907?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6479721860?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/7f13587a59ff3c5350b77bf9334994f91f05d9fc/checks/6479721860/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error[E0308]: mismatched types
   |
  ::: src\librustdoc\html\render\context.rs:75:1
   |
75 |   rustc_data_structures::static_assert_size!(Context<'_>, 144);

For more information about this error, try `rustc --explain E0308`.
[RUSTC-TIMING] rustdoc test:false 4.323
error: could not compile `rustdoc` due to previous error
```

</details>

sunfishcode committed:
Don't test the rustdoc rendering context size on Windows.

This assert is just making sure the size of `Context` doens't grow
unexpectedly, and it's already not being checked on every platform.

`PathBuf` now has a different size on Windows, so adjust this to
avoid checking the size on Windows.

sunfishcode commented:
src/librustdoc/html/render/context.rs had an assert that rustdoc's render context was a certain size. It's already not checked on most platforms, as it's just a guard against the context growing unexpectedly, so I've now added a patch to disable the check on Windows too.

bors commented:
:umbrella: The latest upstream changes (presumably #97433) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

crlf0710 commented:
There's merge conflict now. Needs a rebase.

sunfishcode commented:
r? @joshtriplett

sunfishcode commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

joshtriplett commented:
@bors r+

bors commented:
:pushpin: Commit e620b42644e3404377fa0fae0afd660c6bd77999 has been approved by `joshtriplett`

<!-- @bors r=joshtriplett e620b42644e3404377fa0fae0afd660c6bd77999 -->
<!-- homu: {"type":"Approved","sha":"e620b42644e3404377fa0fae0afd660c6bd77999","approver":"joshtriplett"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/98381#issuecomment-1163241440) for pull requests below priority 1000. This pull request will be tested once the tree is reopened.

klensy commented:
[Merge branch 'master' into main](https://github.com/rust-lang/rust/pull/96869/commits/e620b42644e3404377fa0fae0afd660c6bd77999)
https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy
There was warning for it previously?

sunfishcode commented:
I think I used the github merge-conflict UI and didn't realize that generated a merge commit. I've now fixed that.

r? @joshtriplett

sunfishcode commented:
Ping @joshtriplett; this is already `r+` above, it just needs an approval because of the merge-commit issue which is now fixed.

jyn514 commented:
@bors r=joshtriplett

bors commented:
:pushpin: Commit ddeb936c6d797d47ec8652c885c8a2cceba8386f has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett ddeb936c6d797d47ec8652c885c8a2cceba8386f -->
<!-- homu: {"type":"Approved","sha":"ddeb936c6d797d47ec8652c885c8a2cceba8386f","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ddeb936c6d797d47ec8652c885c8a2cceba8386f with merge 25ea5a36c6e3a4fa1f739538d9e89eb1cd747564...
<!-- homu: {"type":"BuildStarted","head_sha":"ddeb936c6d797d47ec8652c885c8a2cceba8386f","merge_sha":"25ea5a36c6e3a4fa1f739538d9e89eb1cd747564"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7987319538?check_suite_focus=true)
Approved by: joshtriplett
Pushing 25ea5a36c6e3a4fa1f739538d9e89eb1cd747564 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joshtriplett","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7987319538?check_suite_focus=true"},"merge_sha":"25ea5a36c6e3a4fa1f739538d9e89eb1cd747564"} -->

rust-timer commented:
Finished benchmarking commit (25ea5a36c6e3a4fa1f739538d9e89eb1cd747564): [comparison URL](https://perf.rust-lang.org/compare.html?start=87991d5f5d72d6baca490141cb890211ba2f3843&end=25ea5a36c6e3a4fa1f739538d9e89eb1cd747564&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=87991d5f5d72d6baca490141cb890211ba2f3843&end=25ea5a36c6e3a4fa1f739538d9e89eb1cd747564&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=87991d5f5d72d6baca490141cb890211ba2f3843&end=25ea5a36c6e3a4fa1f739538d9e89eb1cd747564&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.3%     | [2.3%, 2.3%] | 1         |
| Regressions ❌ <br /> (secondary) | 4.9%     | [2.6%, 7.2%] | 2         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -2.8%    | [-3.9%, -2.0%] | 4         |
| All ❌✅ (primary) | 2.3%     | [2.3%, 2.3%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=87991d5f5d72d6baca490141cb890211ba2f3843&end=25ea5a36c6e3a4fa1f739538d9e89eb1cd747564&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

RalfJung commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -117,6 +135,14 @@ impl CodePoint {
+    /// It is possible for `bytes` to have valid UTF-8 without this being
+    /// set, such as when we're concatenating `&Wtf8`'s and surrogates become
+    /// paired, as we don't bother to rescan the entire string.
+    is_known_utf8: bool,
```
Adding a field here introduced a subtle bug in PathBuf: https://github.com/rust-lang/rust/issues/124409.
